partiql-lang/src/main/kotlin/org/partiql/lang/eval/PartiqlAstExtensions.kt

```diff
@@ -65,3 +65,115 @@ internal fun PartiqlAst.Expr.getStartingSourceLocationMeta(): SourceLocationMeta
+        override fun visitFromSourceScan(node: PartiqlAst.FromSource.Scan, accumulator: Set<String>): Set<String> {
+            val aliases = listOfNotNull(node.asAlias?.text, node.atAlias?.text, node.byAlias?.text)
+            return accumulator + aliases
+        }
+
+        override fun visitLetBinding(node: PartiqlAst.LetBinding, accumulator: Set<String>): Set<String> {
+            val aliases = listOfNotNull(node.name.text)
+            return accumulator + aliases
+        }
+
+        /** Prevents visitor from recursing into nested select statements */
+        override fun walkExprSelect(node: PartiqlAst.Expr.Select, accumulator: Set<String>): Set<String> {
+            return accumulator
+        }
+    }
+
+    return fold.walkFromSource(selectExpr.from, emptySet())
+        .union(selectExpr.fromLet?.let { fold.walkLet(it, emptySet()) } ?: emptySet())
+}
+
+/** Free variables in an expression.
+ *  A variable (a PartiqlAst.Expr.Id) is free, unless it is in the scope of a same-named alias introduced
+ *  in FROM, LET, or GROUP BY clauses. */
+internal fun PartiqlAst.Expr.freeVariables(): Set<String> {
+    val visitorFold = object : PartiqlAst.VisitorFold<Set<String>>() {
+        // The normal mode of operation is walk through the most AST nodes just carrying the accumulator set along.
+
+        // When a variable reference (a PartiqlAst.Expr.Id) is encountered, add it to the accumulator.
+        override fun walkExprId(node: PartiqlAst.Expr.Id, accumulator: Set<String>): Set<String> {
+            return accumulator + node.name.text
+        }
+
+        // The invariant of each walkExprXxx(node: Xxx, accumulator) call
+        // is to add all *free* variables of `node` expression to `accumulator`.
+        // The default implementations work for all expressions except SELECT.
+
+        // Processing the SELECT involves taking care that variable occurrences bound by FROM, LET, and GROUP BY clauses
+        // do not end up in the result, if they are in the scope of those bindings.
+        override fun walkExprSelect(node: PartiqlAst.Expr.Select, accumulator: Set<String>): Set<String> {
+            var current = emptySet<String>()
+
+            // SELECT clauses are processed in the reverse of evaluation order, since that's how variable scoping extends.
+
+            // Note: in `SELECT e AS x ...`, free variables of `e` get into the result,
+            // but `x` *does not*, due to `x` being a 'symbol', not an 'id' in the AST.
+            // (For reference, 'x' *should not* count as either free or bound because it is not a variable,
+            // it is a spec to create an attribute in a struct.)
+            current += walkProjection(node.project, emptySet())
+
+            node.limit?.let { current += walkExpr(it, emptySet()) }
+            node.offset?.let { current += walkExpr(it, emptySet()) }
+
+            node.order?.let { current += walkOrderBy(it, emptySet()) }
```

Comment:
Not sure if the intention is to call all the free variable function on the normalized AST only or not. For Order by, if you run on a unnormalized ast, the following two are false. On a normalized ast, the following are true. Would be great if you can clarify your intension in the function comments (and the test cases since they are not running on normalized ast). 
```
VarsTestCase("SELECT t.x + 1 as y FROM Tbl t ORDER BY y", setOf("Tbl"))
VarsTestCase("SELECT x + 1 as y FROM Tbl t ORDER BY y",setOf("Tbl", "x"))
```